text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,262 @@
+(Argument name is just convention, will be passed positionally.)
+
+The new optional named argument `background` registers the testbench as a background process when true.
+Processes created through `add_process` are always registered as background processes (except when registering legacy non-async generator functions).
+
+The simulator context has the following properties and methods:
+- `process_type`
+  - Property, `ProcessType`, indicates the type of the current process.
+    This property allows a generic simulation helper function to assert that it's being run under the appropriate process type.
+- `get(expr: Value) -> int`
+- `get(expr: ValueCastable) -> any`
+  - Returns the value of `expr` when awaited.
+    When `expr` is a value-castable, and its `shape()` is a `ShapeCastable`, the value will be converted through the shape's `.from_bits()`.
+    Otherwise, a plain integer is returned.
+- `set(expr: Value, value: ConstLike)`
+- `set(expr: ValueCastable, value: any)`
+  - Set `expr` to `value` when awaited.
+    When `expr` is a value-castable, and its `shape()` is a `ShapeCastable`, the value will be converted through the shape's `.const()`.
+    Otherwise, it must be a const-castable `ValueLike`.
+- `memory_read(instance: MemoryIdentity, address: int)`
+  - Read the value from `address` in `instance` when awaited.
+- `memory_write(instance: MemoryIdentity, address: int, value: int, mask:int = None)`
+  - Write `value` to `address` in `instance` when awaited. If `mask` is given, only the corresponding bits are written.
+    Like `MemoryInstance`, these two functions are an internal interface that will be usually only used via `lib.Memory`.
+    It comes without a stability guarantee.
+- `tick(domain="sync", *, context=None)`
+  - Create a domain trigger object for advancing simulation until the next active edge of the `domain` clock.
+    When an elaboratable is passed to `context`, `domain` will be resolved from its perspective.
+  - If `domain` is asynchronously reset while this is being awaited, `amaranth.sim.AsyncReset` is raised.
+- `delay(interval: float)`
+  - Create a combinable trigger object for advancing simulation by `interval` seconds.
+- `changed(*signals)`
+  - Create a combinable trigger object for advancing simulation until any signal in `signals` changes.
+- `edge(signal, value)`
+  - Create a combinable trigger object for advancing simulation until `signal` is changed to `value`.
+    `signal` must be a 1-bit signal or a 1-bit slice of a signal.
+    `value` is a boolean where true indicates rising edge and false indicates falling edge.
+- `posedge(signal)`
+- `negedge(signal)`
+  - Aliases for `edge(signal, 1)` and `edge(signal, 0)` respectively.
+- `critical()`
+  - Context manager.
+    If the current process is a background process, `async with sim.critical():` makes it a non-background process for the duration of the statement.
+
+The `ProcessType` enum has the following members:
```

Comment:
I don't like the idea that we are giving the ability to *branch* on this rather than *assert* on this. It seems suspiciously powerful for the requested feature.

I think a `@sim.testbench_helper` decorator should work just as well for the purpose of denying its use from `add_process` requested by @wanda-phi:

```
class StreamInterface:
    @sim.testbench_helper
    async def recv(self, sim) -> ConstLike:
        ...
```

Reply:
I think attempting to disallow branching is just going to invite even more cursed hacks if somebody really wants to.

```python
async def is_testbench():
    @sim.testbench_helper
    async def foo():
        return True
    try:
        return await foo()
    except:
        return False
```

Reply:
I'm OK with that because then I want to hear from them and learn why they want it.